objects/src/accounts/delta/vault.rs

```diff
@@ -15,314 +18,466 @@ use crate::{
 /// [AccountVaultDelta] stores the difference between the initial and final account vault states.
 ///
 /// The difference is represented as follows:
-/// - added_assets: a vector of assets that were added to the account vault.
-/// - removed_assets: a vector of assets that were removed from the account vault.
+/// - fungible: a binary tree map of fungible asset balance changes in the account vault.
+/// - non_fungible: a binary tree map of non-fungible assets that were added to or removed from the
+///   account vault.
 #[derive(Clone, Debug, Default, PartialEq, Eq)]
 pub struct AccountVaultDelta {
-    pub added_assets: Vec<Asset>,
-    pub removed_assets: Vec<Asset>,
+    fungible: FungibleAssetDelta,
+    non_fungible: NonFungibleAssetDelta,
 }
 
 impl AccountVaultDelta {
-    /// Creates an empty [AccountVaultDelta].
-    pub fn empty() -> Self {
-        Self::default()
+    /// Validates and creates an [AccountVaultDelta] with the given fungible and non-fungible asset
+    /// deltas.
+    ///
+    /// # Errors
+    /// Returns an error if the delta does not pass the validation.
+    pub fn new(
+        fungible: FungibleAssetDelta,
+        non_fungible: NonFungibleAssetDelta,
+    ) -> Result<Self, AccountDeltaError> {
+        let delta = Self { fungible, non_fungible };
+
+        delta.validate()?;
```

Comment:
Why do we need this call here? `fungible` and `non_fungible` deltas should already be validated, right?

Reply:
No, we still have validation for fungible assets delta (validation of faucet ID).

Reply:
But fungible asset delta is passed in as `FungibleAssetDelta` - which should already be in a valid sate. Or am I missing something?

Reply:
Ah, you're right, they are validated during creation! I will remove the validation here, thank you!